src/main/java/org/prebid/server/bidder/tappx/TappxBidder.java

```diff
@@ -107,22 +109,30 @@ private String buildEndpointUrl(ExtImpTappx extImpTappx, Integer test) {
             throw new PreBidException("Tappx tappxkey undefined");
         }
 
-        String url = String.format("%s%s/%s?tappxkey=%s", endpointUrl, host, endpoint, tappxkey);
-        if (test != null && test == 0) {
-            int t = (int) System.nanoTime();
-            url += "&ts=" + t;
-        }
-
-        url += "&v=" + VERSION;
-        url += "&type_cnn=" + TYPE_CNN;
+        return buildUrl(endpointUrl, host, endpoint, tappxkey, test);
+    }
 
+    private String buildUrl(String endpointUrl, String host, String endpoint, String tappxkey, Integer test) {
         try {
-            HttpUtil.validateUrl(url);
-        } catch (IllegalArgumentException e) {
-            throw new PreBidException("Not valid url: " + url, e);
-        }
+            final URIBuilder uriBuilder = new URIBuilder(endpointUrl + host);
 
-        return url;
+            if (!host.contains(endpoint)) {
+                uriBuilder.setPath(endpoint);
+            }
+
+            uriBuilder.addParameter("tappxkey", tappxkey);
+
+            if (test != null && test == 0) {
+                int t = (int) System.nanoTime();
+                uriBuilder.addParameter("ts", String.valueOf(t));
+            }
+
+            uriBuilder.addParameter("v", VERSION);
+            uriBuilder.addParameter("type_cnn", TYPE_CNN);
+            return uriBuilder.build().toString();
+        } catch (URISyntaxException e) {
+            throw new PreBidException("Built invalid URL: ", e);
```

Comment:
throw new PreBidException(String.format("Failed to build endpoint URL: %s", e.getMessage()));

Reply:
Done